### PR TITLE
fix(encoding): error handling for TypeErrors in v0.5 encoding for wrong span meta types

### PR DESCRIPTION
**Pull request by Yun-Kim**
## Description
Unblocks #4404.

This PR adds exception handling for `TypeError` in the v0.5 encoder specifically for `_pack_str()`. Previously, if non-text types were provided, the v0.5 encoder would only silently pass the `TypeError` into stderr, and not loudly propagate the exception back to top-level like we should expect it to. This resulted in tests for setting non-text span tags (i.e. lists, dictionaries, numbers) with the v0.5 encoder failing to catch any errors.

## Checklist
- [ ] Followed the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) when writing a release note.
- [ ] Add additional sections for `feat` and `fix` pull requests.
- [ ] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [x] Title is accurate.
- [x] Description motivates each change.
- [x] No unnecessary changes were introduced in this PR.
- [x] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Release note has been added and follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines), or else `changelog/no-changelog` label added.
- [x] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.

### CODE REVIEW

File: ddtrace/internal/_encoding.pyx [hunk 1]
```diff
@@ -166,22 +166,27 @@ cdef class StringTable(object):
...
 
         if string is None:
             return 0
 
-        if PyDict_Contains(self._table, string):
-            return PyLong_AsLong(<object>PyDict_GetItem(self._table, string))
+        ret = PyDict_Contains(self._table, string)
+        if ret == -1: return ret
+        if ret:
+            ret = PyLong_AsLong(<object>PyDict_GetItem(self._table, string))
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/_encoding.pyx:179]:
```suggestion
            return PyLong_AsLong(<object>PyDict_GetItem(self._table, string))
```

we can revert this since if `PyLong_AsLong` returns `-1` then we want to return anyway.

File: ddtrace/internal/_encoding.pyx [hunk 2]
```diff
@@ -166,22 +166,27 @@ cdef class StringTable(object):
...
+            if ret != 0: return ret
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/_encoding.pyx:180]: I think this is actually wrong. If `PyLong_AsLong` returns 0 then we'll continue even though the value is in the table already.

File: releasenotes/notes/v05-encoding-pack-str-error-handle-d11960b72472af7c.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
+  - |
+    tracing: This fix resolves an issue where using the v0.5 encoding endpoint did not raise any ``TypeError``
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/v05-encoding-pack-str-error-handle-d11960b72472af7c.yaml:4]:
```suggestion
    tracing: Resolves an issue where using the v0.5 encoding endpoint did not raise any ``TypeError``
```

is "This fix" redundant given it's in the fixes section?

File: releasenotes/notes/v05-encoding-pack-str-error-handle-d11960b72472af7c.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+    if non-text span tags were set.
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/v05-encoding-pack-str-error-handle-d11960b72472af7c.yaml:5]: Maybe worth rewording to be "v0.5 encoding endpoint errors were output to stderr instead of to a logger..."? The semantics of a user's application isn't affected by this change. I believe the error would have been caught and logged still, it just got printed to stderr directly instead of to a logger.

**Review by Kyle-Verhoog — COMMENTED** [review]
Can we add a regression test?

File: tests/integration/test_integration.py [hunk 5]
```diff
@@ -869,3 +868,47 @@ def test_no_warnings():
...
+@pytest.mark.parametrize(
+    "meta",
+    [
+        ({"env": "my-env", "tag1": "some_str_1", "tag2": "some_str_2", "tag3": [1, 2, 3]}),
+        ({"env": "test-env", b"tag1": {"wrong_type": True}, b"tag2": "some_str_2", b"tag3": "some_str_3"}),
+        ({"env": "my-test-env", u"😐": "some_str_1", b"tag2": "some_str_2", "unicode": 12345}),
+    ],
+)
+@allencodings
+def test_trace_with_wrong_meta_types(encoding, meta, monkeypatch):
```
Comment (Kyle-Verhoog) [inline tests/integration/test_integration.py:882]: could we just reuse the snapshot tests below?

File: releasenotes/notes/v05-encoding-pack-str-error-handle-d11960b72472af7c.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
+  - |
+    tracing: This resolves an issue where v0.5 encoding endpoint errors were output to ``stderr`` instead of
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/v05-encoding-pack-str-error-handle-d11960b72472af7c.yaml:4]:
```suggestion
    tracing: Resolves an issue where v0.5 encoding endpoint errors were output to ``stderr`` instead of
```

File: ddtrace/internal/_encoding.pyx [hunk 7]
```diff
@@ -166,22 +166,26 @@ cdef class StringTable(object):
...
-        if PyDict_Contains(self._table, string):
+        ret = PyDict_Contains(self._table, string)
+        if ret == -1: return ret
+        if ret:
             return PyLong_AsLong(<object>PyDict_GetItem(self._table, string))
 
         _id = self._next_id
-        PyDict_SetItem(self._table, string, PyLong_FromLong(_id))
+        ret = PyDict_SetItem(self._table, string, PyLong_FromLong(_id))
+        if ret != 0: return ret
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/_encoding.pyx:183]:
```suggestion
        if ret == -1: return ret
```

(although seems like != 0 is fine too https://docs.python.org/3/c-api/dict.html#c.PyDict_SetItem)

but we already do `ret == -1` above so worth being consistent imo

**Review by Kyle-Verhoog — COMMENTED** [review]
a few nits

File: releasenotes/notes/v05-encoding-pack-str-error-handle-d11960b72472af7c.yaml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
+  - |
+    tracing: Resolves an issue where v0.5 encoding endpoint errors were output to ``stderr`` instead of
+    to a logger if non-text span tags were set.
```
Comment (mabdinur) [inline releasenotes/notes/v05-encoding-pack-str-error-handle-d11960b72472af7c.yaml:5]: Could we simplify the release note it's a bit wordy/confusing. Maybe something like this would be better `tracing: Ensure encoding errors are logged`

Also does this change require a release note? MsgPack encoders are an internal component and this fix is for an encoder that is disabled by default.

File: tests/integration/test_integration.py [hunk 9]
```diff
@@ -208,7 +208,6 @@ def test_child_spans(encoding, monkeypatch):
             spans.append(t.trace("op"))
         for s in spans:
             s.finish()
-
```
Comment (mabdinur) [inline tests/integration/test_integration.py:211]: nit: we can revert this change. It's not related to this PR

File: releasenotes/notes/v05-encoding-pack-str-error-handle-d11960b72472af7c.yaml (hunk 8 above)
Comment (Yun-Kim) [reply releasenotes/notes/v05-encoding-pack-str-error-handle-d11960b72472af7c.yaml]: Simplified the release note, but I think this change does require a release note as users previously wouldn't have seen encoding errors from their logs, they would have had to dig into their application's stderr directly.

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]
